File: Backend/src/main/java/com/spring/fullstacktodo/service/TaskService.java
package com.spring.fullstacktodo.service;

import com.spring.fullstacktodo.dto.TaskRequestDTO;
import com.spring.fullstacktodo.dto.TaskResponseDTO;
import com.spring.fullstacktodo.exception.TaskNotFoundException;
import com.spring.fullstacktodo.mapper.TaskMapper;
import com.spring.fullstacktodo.model.Task;
import com.spring.fullstacktodo.model.User;
import com.spring.fullstacktodo.repository.TaskRepo;
import com.spring.fullstacktodo.repository.UserRepo;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class TaskService {
    private final TaskRepo taskRepo;
    private final TaskMapper taskMapper;
    private final UserRepo userRepo;

    // Create a new task (scoped to user)
    public TaskResponseDTO createTask(TaskRequestDTO taskRequestDTO, Long userId) {
        Task task = taskMapper.toEntity(taskRequestDTO);
        User user = userRepo.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userId));
        task.setUser(user);
        Task savedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(savedTask);
    }

    // Get all tasks for a user
    public List<TaskResponseDTO> getAllTasks(Long userId) {
        List<Task> tasks = taskRepo.findAllByUserIdOrderByCreatedAtDesc(userId);
        return taskMapper.toResponseDtoList(tasks);
    }

    // Get task by id (scoped to user)
    public Optional<TaskResponseDTO> getTaskById(Long id, Long userId) {
        return taskRepo.findByIdAndUserId(id, userId)
                .map(taskMapper::toResponseDto);
    }

    // Update task
    public TaskResponseDTO updateTask(Long id, TaskRequestDTO taskRequestDTO, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        taskMapper.updateEntityFromDto(taskRequestDTO, task);
        Task updatedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(updatedTask);
    }

    // Delete task
    public void deleteTask(Long id, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        taskRepo.delete(task);
    }

    // Delete all tasks for a user
    public void deleteAllTasks(Long userId) {
        taskRepo.deleteAllByUserId(userId);
    }

    // Mark task as completed
    public TaskResponseDTO markTaskAsCompleted(Long id, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        task.setStatus(Task.TaskStatus.DONE);
        Task updatedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(updatedTask);
    }

    // Mark task as in progress
    public TaskResponseDTO markTaskAsInProgress(Long id, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        task.setStatus(Task.TaskStatus.IN_PROGRESS);
        Task updatedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(updatedTask);
    }

    // Mark task as to do
    public TaskResponseDTO markTaskAsToDo(Long id, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        task.setStatus(Task.TaskStatus.TODO);
        Task updatedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(updatedTask);
    }

    // Mark task as urgent
    public TaskResponseDTO markTaskAsUrgent(Long id, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        task.setPriority(Task.TaskPriority.URGENT);
        Task updatedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(updatedTask);
    }

    // Mark task as high
    public TaskResponseDTO markTaskAsHigh(Long id, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        task.setPriority(Task.TaskPriority.HIGH);
        Task updatedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(updatedTask);
    }

    // Mark task as medium
    public TaskResponseDTO markTaskAsMedium(Long id, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        task.setPriority(Task.TaskPriority.MEDIUM);
        Task updatedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(updatedTask);
    }

    // Mark task as low
    public TaskResponseDTO markTaskAsLow(Long id, Long userId) {
        Task task = taskRepo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new TaskNotFoundException(id));
        task.setPriority(Task.TaskPriority.LOW);
        Task updatedTask = taskRepo.save(task);
        return taskMapper.toResponseDto(updatedTask);
    }

    // Get task by status
    public List<TaskResponseDTO> getTasksByStatus(Task.TaskStatus status, Long userId) {
        List<Task> tasks = taskRepo.findByUserIdAndStatus(userId, status);
        return taskMapper.toResponseDtoList(tasks);
    }

    // Get tasks by priority
    public List<TaskResponseDTO> getTasksByPriority(Task.TaskPriority priority, Long userId) {
        List<Task> tasks = taskRepo.findByUserIdAndPriority(userId, priority);
        return taskMapper.toResponseDtoList(tasks);
    }

    // Search tasks by title
    public List<TaskResponseDTO> searchTasksByTitle(String title, Long userId) {
        List<Task> tasks = taskRepo.findByUserIdAndTitleContainingIgnoreCase(userId, title);
        return taskMapper.toResponseDtoList(tasks);
    }

    // Get all tasks ordered by priority
    public List<TaskResponseDTO> getAllTasksOrderedByPriority(Long userId) {
        List<Task> tasks = taskRepo.findAllByUserIdOrderByPriorityDesc(userId);
        return taskMapper.toResponseDtoList(tasks);
    }

    // Get all tasks ordered by created date
    public List<TaskResponseDTO> getAllTasksOrderedByDate(Long userId) {
        List<Task> tasks = taskRepo.findAllByUserIdOrderByCreatedAtDesc(userId);
        return taskMapper.toResponseDtoList(tasks);
    }
}

File: Frontend/src/services/TaskService.js

import axios from 'axios';
import { API_BASE_URL } from '../utils/Constants';

const taskService = {
  // Get all tasks
  getAllTasks: async () => {
    const response = await axios.get(API_BASE_URL);
    return response.data;
  },

  // Get task by ID
  getTaskById: async (id) => {
    const response = await axios.get(`${API_BASE_URL}/${id}`);
    return response.data;
  },

  // Create new task
  createTask: async (task) => {
    const response = await axios.post(API_BASE_URL, task);
    return response.data;
  },

  // Update task
  updateTask: async (id, task) => {
    const response = await axios.put(`${API_BASE_URL}/${id}`, task);
    return response.data;
  },

  // Delete task
  deleteTask: async (id) => {
    await axios.delete(`${API_BASE_URL}/${id}`);
  },

  // Mark task as completed
  markAsCompleted: async (id) => {
    const response = await axios.patch(`${API_BASE_URL}/${id}/complete`);
    return response.data;
  },

  // Mark task as in progress
  markAsInProgress: async (id) => {
    const response = await axios.patch(`${API_BASE_URL}/${id}/in-progress`);
    return response.data;
  },

  // Mark task as todo
  markAsTodo: async (id) => {
    const response = await axios.patch(`${API_BASE_URL}/${id}/todo`);
    return response.data;
  },

  // Get tasks by status
  getTasksByStatus: async (status) => {
    const response = await axios.get(`${API_BASE_URL}/status/${status}`);
    return response.data;
  },

  // Get tasks by priority
  getTasksByPriority: async (priority) => {
    const response = await axios.get(`${API_BASE_URL}/priority/${priority}`);
    return response.data;
  },

  // Search tasks by title
  searchTasks: async (title) => {
    const response = await axios.get(`${API_BASE_URL}/search`, {
      params: { title }
    });
    return response.data;
  },

  // Get tasks ordered by priority
  getTasksOrderedByPriority: async () => {
    const response = await axios.get(`${API_BASE_URL}/order/priority`);
    return response.data;
  },

  // Get tasks ordered by date
  getTasksOrderedByDate: async () => {
    const response = await axios.get(`${API_BASE_URL}/order/date`);
    return response.data;
  }
};

export default taskService;